### DISCUSSION TITLE
Linking against a library that links against other libraries

### DISCUSSION THREAD

**Discussion by Epsylene** [premake/premake-core#1816] [Q&A] [answered]
I am trying to write a premake file for an app that links against a custom library, which in turn links against GLFW. The test premake file looks like this:

```lua
workspace "HelloWorld"
    configurations { "Debug", "Release" }

project "Lib"
    kind "SharedLib"
    language "C++"
    targetdir "bin"

    files { "src/**.h", "src/**.cpp" }

    includedirs "vendor/glfw/include"
    libdirs "vendor/glfw/lib"
    links { "glfw3", "opengl32" }

project "App"
    kind "ConsoleApp"
    language "C++"
    targetdir "bin"
    staticruntime "On"

    files { "app/**.h", "app/**.cpp" }

    includedirs { "src", "vendor/glfw/include" }

    libdirs "vendor/glfw/lib"
    links { "Lib", "glfw3", "gdi32" }
```

The library directory contains a single file with a function that calls some GLFW test code, and the app is comprised of a single file that calls this function; [here is a ZIP file](https://github.com/premake/premake-core/files/8088192/files.zip) containing them both. GLFW was pre-built as a static library here.

The problem is that if I don't link `App` against `glfw3` and `gdi32`, I get a bunch of "undefined reference" errors from GLFW and libgdi32. I don't understand why this is, even though I am linking against my library, which links against those libraries; since my library DLL is placed along my executable in the `bin` directory, there should be no problem, right ? If possible, I would like to allow users of my library to only have to link against my library, without caring about its dependencies (is this realistic ? I still lack a lot of understanding when it comes to linking and managing libraries and their dependencies).

Excuse me if this is not the right place to ask these kind of questions, I couldn't find (or understand) a clear answer to this problem elsewhere.

**Comment by samsinsane** [comment, marked as answer]
Your `Lib` library doesn't contain any code - functions in headers like `hello.h` aren't compiled in the library but in the file that `#include`s them, in this case `app.cpp`. You need to put the contents of `notmain()` into `hello.cpp` and then export it like this:
```c++
// Windows
__declspec(dllexport) int notmain();

// Not-Windows
__attribute__((visibility("default"))) int notmain();
```

You can see how GLFW does it [here](https://github.com/glfw/glfw/blob/97da62a027794d9ff0f4512268cb9a73a8fb5073/include/GLFW/glfw3.h#L258-L273) and it is used [here](https://github.com/glfw/glfw/blob/97da62a027794d9ff0f4512268cb9a73a8fb5073/include/GLFW/glfw3.h#L2132).

**Reply by Epsylene** [reply]
I see, thanks for the reply !
